### PR TITLE
[WINDUPRULE-791] Add rule for detecting usage of sun.misc base64 encoders

### PR DESCRIPTION
**Pull request by jmle**
See [Jira ticket](https://issues.redhat.com/browse/WINDUPRULE-791).

### CODE REVIEW

**Comment by gnu-andrew** [comment]
This looks right to me. I was going to do this one, because it seemed pretty simple, but thanks for taking it on.

**Comment by jmle** [comment]
> This looks right to me. I was going to do this one, because it seemed pretty simple, but thanks for taking it on.

@gnu-andrew no problem! It would be great if you could take a look at [this PR](https://github.com/windup/windup-rulesets/pull/637) too, since I wasn't very sure about the hint wording.

File: rules-reviewed/openjdk11/openjdk8/tests/java-removals.windup.test.xml [hunk 1]
```diff
@@ -54,6 +55,18 @@
...
+                    </iterable-filter>
+                </not>
+            </when>
+            <perform>
+                <fail message="[java-removals] `java-removals-00101-test failed"/>
+            </perform>
+        </rule>
+    </rules>
+</ruleset>
+        </ruletest>
```
Comment (carlosthe19916) [inline rules-reviewed/openjdk11/openjdk8/tests/java-removals.windup.test.xml:72]: These 2 last closing tags are not aligned. Could you please format them?

File: rules-reviewed/openjdk11/openjdk8/tests/data/java-removals/Base64Removals.java [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+        BASE64Decoder base64Decoder = new BASE64Decoder();
+
+        try {
+            base64Decoder.decodeBuffer("inputString");
+            base64Encoder.encodeBuffer(new byte[]{'a', 'b'});
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+    }
+}
```
Comment (carlosthe19916) [inline rules-reviewed/openjdk11/openjdk8/tests/data/java-removals/Base64Removals.java:20]: If I replace this file content with:
```
package com.test.project;

import sun.misc.*;

import java.io.IOException;

public class Base64Removals {
    public static void main(String[] args) {
        BASE64Encoder base64Encoder = new BASE64Encoder();

        try {
            base64Decoder.decodeBuffer("inputString");
            base64Encoder.encodeBuffer(new byte[]{'a', 'b'});
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}
```

Then no Hint is being fired. I doubt this has to to with the Rule created in this PR but I'd like to understand/know where the issue is and how to fix it.

**Review by carlosthe19916 — COMMENTED** [review]
@jmle The PR and Rule definition looks good to me. However, I've found an edge case where a Valid Java Class doesn't generate an appropriate Hint. 

Steps to reproduce the issue:
- Replace the content of the file `rules-reviewed/openjdk11/openjdk8/tests/data/java-removals/Base64Removals.java` by:

```
package com.test.project;

import sun.misc.*;

import java.io.IOException;

public class Base64Removals {
    public static void main(String[] args) {
        BASE64Encoder base64Encoder = new BASE64Encoder();

        try {
            base64Decoder.decodeBuffer("inputString");
            base64Encoder.encodeBuffer(new byte[]{'a', 'b'});
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}
```

- Execute:
```
mvn clean test -DrunTestsMatching=java-removals
```
- Open the static report generated at `target/WindupRulesTests/windupgraph_UUID/reports/index.html` and go to the `Issues Tab`
- There is no Hint generated. I expect a hint to be generated since the Java class is a valid one and contains the line `BASE64Encoder base64Encoder = new BASE64Encoder()`

NOTE: the trick to generate the bug was to use an `import` that uses the `*` character. and also include only either BASE64Encoder or BASE64Decoder in the class.

File: rules-reviewed/openjdk11/openjdk8/java-removals.windup.xml [hunk 3]
```diff
@@ -79,5 +79,23 @@
                 </hint>
             </perform>
         </rule>
+        <rule id="java-removals-00101">
+            <when>
+                <or>
+                    <javaclass references="sun.misc.BASE64Encoder"/>
+                    <javaclass references="sun.misc.BASE64Decoder"/>
+                </or>
```
Comment (mrizzi) [inline rules-reviewed/openjdk11/openjdk8/java-removals.windup.xml:87]: I think it would be great to try to have just one condition with a `where` parameter to "inject" `Encoder` or `Decoder` and use the same also to drive the hint's message below.

**Review by mrizzi — COMMENTED** [review]
Added a comment to suggest to leverage a `where` parameter.

**Comment by carlosthe19916** [comment]
@jmle @mrizzi as discussed internally, I've created a Ticket here https://issues.redhat.com/browse/WINDUPRULE-832 so we can work on the bug in a separate PR.

I've tested this PR again and it works fine, even in the Edge case I reported previously in this PR. I've also found that lines like

```
BASE64Encoder base64Encoder = new BASE64Encoder();
```

Are going to generate twice the same hint. Eg.
Screenshot from 2022-03-25 10-43-52

And this "double hint" was, as discussed internally, acceptable. So I will approve this PR!

**Review by carlosthe19916 — APPROVED** [review]
